### ISSUE TITLE
Multi domain authentication, same forest

### ISSUE THREAD

**Issue by aabuasad** [Adldap2/Adldap2-Laravel#279]
- Laravel Version: 5.4
- Adldap2-Laravel Version: dev-master
- PHP Version: 7.1.1

### Description:
Hello and thanks for this amazing package, and the up to the point walk through, i have managed to get logins working with LDAP, but my issue is that i have users from 2 domains that would use my website, and both domains are under the same forest. on the login form users are requested to provide their email and password.

I have created two connections under adldap.php, named them alpha and beta, now under adldap_auth, how can i have it check on both connections? or better have it check connection alpha if the email ends with alpha.com, and beta if the email ends with beta.com?

**Comment by stevebauman** [comment]
Hi @aabuasad, thanks for moving this issue.

Unfortunately the Adldap2 auth driver only supports authenticating against one domain.

However, you can perform this by determining which domain the user wants to authenticate to using a select box and allowing the user to select a domain, or by checking their email address.

Here's how you can do it:

```blade
<!-- Inside your login form -->

<input name="email" type="text" />

<input name="password" type="password" />

<select name="domain">
    <option value="alpha">Alpha</option>
   <option value="beta">Beta</option>
</select>
``` 

```php
// Inside your LoginController.php

public function login(Request $request)
{
    if ($request->domain == 'alpha') {
        Config::set('adldap_auth.connection', 'alpha');
    } else {
        Config::set('adldap_auth.connection', 'beta');
    }

    if (Auth::attempt($request->only('email', 'password'))) {
        return redirect()->to('/home');
    }

    return back();
}
```

Or by their email:

```php
// Inside your LoginController.php

public function login(Request $request)
{
    if (Str::contains($request->email, '@alpha.com')) {
        Config::set('adldap_auth.connection', 'alpha');
    } else {
        Config::set('adldap_auth.connection', 'beta');
    }

    if (Auth::attempt($request->only('email', 'password'))) {
        return redirect()->to('/home');
    }

    return back();
}
```

Can you give this a shot and see if it works for you?

**Comment by aabuasad** [comment]
Thanks for your time Steve, i'm still a newbie here myself, been learning Laravel for a month or so. now, i have followed your quick start guide and use the login skeleton provided by Laravel, authentication wise i haven't done anything beyond your guide, than enabling some rules in adldap_auth. with that being said, can you tell me how can i incorporate your solution above?

**Comment by stevebauman** [comment]
No problem, we all start off learning at some point!

I'm assuming you know HTML as well as thorough PHP knowledge with the below instructions.

If you have the login skeleton provided by Laravel as shown in the quick start guide, here's how you can implement it.

Edit the file `resources/views/auth/login.blade.php`, and add the domain select box inside the login form:

```php
<select name="domain">
    <option value="alpha">Alpha</option>
    <option value="beta">Beta</option>
</select>
```

Then inside your `app/Http/Controllers/Auth/LoginController.php`, use the following:

> **Note**: All I've done is add the protected method `attemptLogin`

```php
<?php

namespace App\Http\Controllers\Auth;

use Illuminate\Http\Request;
use App\Http\Controllers\Controller;
use Illuminate\Support\Facades\Config;
use Illuminate\Foundation\Auth\AuthenticatesUsers;

class LoginController extends Controller
{
    /*
    |--------------------------------------------------------------------------
    | Login Controller
    |--------------------------------------------------------------------------
    |
    | This controller handles authenticating users for the application and
    | redirecting them to your home screen. The controller uses a trait
    | to conveniently provide its functionality to your applications.
    |
    */

    use AuthenticatesUsers;

    /**
     * Where to redirect users after login.
     *
     * @var string
     */
    protected $redirectTo = '/home';

    /**
     * Create a new controller instance.
     *
     * @return void
     */
    public function __construct()
    {
        $this->middleware('guest', ['except' => 'logout']);
    }

    /**
     * Attempt to log the user into the application.
     *
     * @param  \Illuminate\Http\Request  $request
     * @return bool
     */
    protected function attemptLogin(Request $request)
    {
        if ($request->domain == 'alpha') {
            Config::set('adldap_auth.connection', 'alpha');
        } else {
            Config::set('adldap_auth.connection', 'beta');
        }

        return $this->guard()->attempt(
            $this->credentials($request), $request->has('remember')
        );
    }
}
```

**Comment by aabuasad** [comment]
Many Thanks Steve!!!!!! i know my way around PHP and HTML, but i'm still new to Laravel, and your solution here made things super easy and beautiful! thanks again, have a nice weekend.

**Comment by riyazpt** [comment]
Hi ,
As you mentioned  I have changed adldap_auth.php  ` 'connection' => env('ADLDAP_CONNECTION', `'alpha'),

and adldap.php   'connections' => [

        'alpha' => [
]
]

But i am getting  following error 


(1/1) AdldapExceptionThe connection provider 'default' does not exist.
--
in Adldap.php (line 75)

Please help

**Comment by riyazpt** [comment]
Also tried 

        Config::set('adldap_auth.connection', 'alpha');
   not setting to alpha connection .Always it checks for default connection if I  change Default Connection An Error occurs 



(1/1) AdldapExceptionThe connection provider 'default' does not exist.
--
in Adldap.php (line 75)

**Comment by zmiKaminski** [comment]
I'm also having this issue.  I implemented this exact technique in a Laravel 5.2 application using `"adldap2/adldap2-laravel": "2.1.*"` and it works great.  Now I am trying to implement this in Laravel 5.5 with `"adldap2/adldap2-laravel": "^4.0",`

To elaborate, in my `adldap.php` I have an array of connections, none of them are titled "default".
```
return [ 
    'connections' => [
        'test1' => [..],
        'test2' => [..],
        'test3' => [..],
    ]
];
```

And in my `adldap_auth.php` config:
```
    'connection' => 'test1'
```

And I receive the same error that the previous comment received.
`AdldapException The connection provider 'default' does not exist.`

If I rename `'test1'` to `'default'` in `adldap.php` it works.

No matter what it keeps looking for a default connection.  I've cleared my config cache *a lot* and reset the PHP OpCache.

Willing to provide any help necessary, this package is great and you're great.

**Comment by zmiKaminski** [comment]
More information:  It appears 'default' is used inside of `\Adldap\Laravel\Resolvers\UserResolver`.  I was able to "hack" around this by overriding this file in my `composer.json` using [this technique](https://stackoverflow.com/questions/28104574/strategy-to-override-a-class-in-a-library-installed-with-composer) and setting the connection from the config for now.  

I'm not sure how this affects the rest of the library, probably not in a good way.

**Comment by stevebauman** [comment]
Hey @zmiKaminski, this is a bug. Going to look into this now, thanks!

**Comment by stevebauman** [comment]
Hi @zmiKaminski, can you give it another go with `dev-master`? Should be fixed, thanks!

**Comment by zmiKaminski** [comment]
@stevebauman I gave it a shot and it's working great.  Thank you very much!

**Comment by stevebauman** [comment]
Great, thanks for the validation @zmiKaminski! Much appreciated. Will be releasing with some tests shortly :)

**Comment by stevebauman** [comment]
Released fix in `v4.0.7`.
